<img src="https://www.python.org/static/img/python-logo@2x.png" style="background-color:#1E415E; float:left"/>

# Presentación del lenguaje

Python es un lenguaje de programación creado por [Guido van Rossum](https://es.wikipedia.org/wiki/Guido_van_Rossum) a principios de los 90s, quién se inspiro en el grupo de humoristas británicos [Monty Python](https://es.wikipedia.org/wiki/Monty_Python) para elegir el nombre.

Este lenguaje, a diferencia de la mayor parte de los lenguajes que le preceden, es muy sencillo de leer y escribir: La persona que lo utiliza puede concentrarse en lo que quiere que el computador haga, complicandose muy poco en cuanto a cómo estructurar las instrucciones para que el computador pueda seguirlas. En otras palabras, Python tiene una estructura sintáctica simple y clara.

La simplicidad de Python permitió este se convirtiera en uno de los lenguajes de programación más populares hoy en dia, permitiendo que personas sin un gran dominio técnico pudan crear programas computacionales en pocas líneas, fáciles de leer, y que cumplieran con el trabajo requerido. Para una persona que sepa leer en inglés, se volverá muy sencillo leer un programa escrito en Python. De todos modos, si no sabe inglés no se preocupe, porque la cantidad de palabras que utiliza Python es tan pequeña, que las aprenderá brevemente.

## Plataforma
Este tutorial, al igual que el resto de los talleres de la escuela, serán desarrollados en formato Jupyter Notebbok sobre la plataforma [lab.quantum-computing.ibm.com](https://lab.quantum-computing.ibm.com)

# Conceptos básicos

En Python, cada línea que usted escriba tiene sentido propio. Por ejemplo, puede escribir operaciones matemáticas, y al correrlas retornan el resultado de esa operación.

Para correr una celda, en el caso de este _notebook_, puede seleccionar la celda a correr con el _mouse_ y luego presionar `Shift + Enter`.

In [1]:
2 + 3

5

En el caso de que tenga varias líneas dentro de una celda, sólo se mostrará automaticamente el resultado de la última línea

In [2]:
1 + 2
4 - 6
2 * 4
99 / 3

33.0

# Variables y comentarios

Una __variable__ es un objeto que tiene nombre y guarda un valor o conjunto de valores. Puede crearlas escribiendo el nombre deseado, y utilizando el operador de asignación `=` para otorgale el valor deseado.

Los nombres de variables pueden contener mayúsculas, minúsculas, guiones bajos, e incluso números en su interior, pero __no pueden contener__ guiones normales ni otros caracteres especiales. Además, los nombres no pueden comenzar con un número.

Python ignorará todo texto que venga en una línea seguido de un símbolo `#`. A este texto se le llama __comentario__, y es útil para dejar mensajes a los humanos que lean su código. Un uso típico de los comentarios es explicar lo que hace una porción de código.

In [3]:
# Soy un comentario! =)

# Aquí declaramos dos variables
mi_variable = 4
mi_variable_2 = 5

# Si sumo los valores de mis variables debería obtener 9
mi_variable + mi_variable_2

9

## Tipos de variable

Para el contenido de este curso sólo nos preocuparemos de 5 tipos de variable fundamentales:
1.  Enteros (`int`)
2.  Reales  (`float`)
3.  Complejos (`complex`)
4.  Lógicas o Booleanas (`bool`)
5.  Cadenas de texto o _Strings_ (`str`)

y podemos averiguar el tipo de una variable utilizando la función de python, `type`.


### Enteros
Los enteros son numeros que trabajamos sin parte decimal, como por ejemplo `4` o `-110`

In [4]:
type(4)

int

### Reales

Si escribimos un número con parte decimal, este será un real.

In [5]:
type(4.0) # También se puede escribir como '4.' y se asume que todo lo que viene despues son ceros

float

In [6]:
4 / 3

1.3333333333333333

In [7]:
4 / 2 # El resultado es un real, incluso si los numeros son divisibles de forma exacta

2.0

### Complejos

Los complejos son numeros que tienen una parte real y otra imaginaria, dónde la parte imaginaria es un numero real multiplicado por la unidad imaginaria, $i$.

En python, la unidad imaginaria se denota con una letra `j`, y esta acompaña a la parte imaginaria de un numero sin usar el simbolo de multiplicación.

In [8]:
1 + 2j # Este es un numero con parte real 1 y parte imaginaria 2.

(1+2j)

In [9]:
cero_complejo = 0j # Este es un numero con partes real e imaginaria cero

In [10]:
type(cero_complejo)

complex

In [11]:
cero_complejo + 2 # Esto es un numero complejo con parte real 2 y parte imaginaria 0

(2+0j)

Para que se entienda como la unidad imaginaria, `j` debe estar junto a un número y sin operaciones de por medio. Si por el contrario se tiene una expresión como
```python
1 + j
```
o
```python
2 - 4*j
```
se entenderá que está operando con una variable llamada `j`, la que deberá estar previamente definida por usted.

### Lógicas

Son variables que tienen un valor de verdad verdadero (`True`) o falso (`False`). La forma más común de obtenerlas es mediante comparaciones de otras variables.

In [12]:
mi_variable_logica = 3 >= 4 # 'mi_variable_logica' tiene el valor de verdad de "3 es mayor o igual a 4"

mi_variable_logica

False

Las comparaciones numéricas más comunes son de igualdad (`==`), mayor o menor estricto (`>`, `<`), mayor o igual (`>=`), menor o igual (`<=`).

#### Operadores lógicos

Existen operadores que actúan sobre variables lógicas. Los más importantes son:

In [13]:
# Negación (actúa sólo sobre una variable, negándola)
not mi_variable_logica

True

In [14]:
# Operador 'y'. Verdadero sólo si sus dos argumentos son verdaderos
mi_variable_logica and True

False

In [15]:
# Operador 'o'. Verdadero si al menos uno de sus argumentos es verdadero
mi_variable_logica or True

True

### _Strings_

Las cadenas de texto o _strings_, son variables que almacenan un mensaje alfanumérico, y se definen delimitandolas entre comillar simples (\').

In [16]:
msj = 'Soy una cadena de texto! Puedo contener números y símbolos estándares: 12312312 !"#%$&/'

msj

'Soy una cadena de texto! Puedo contener números y símbolos estándares: 12312312 !"#%$&/'

In [17]:
type(msj)

str

# Errores

Como en cualquier lenguaje, siempre es posible escribir algún mensaje que no cumpla sus reglas, dejando al computador en una situación en la que no puede interpretar lo que se le pidió. También es posible ordenar al computador que realice una acción que por diversos motivos puede no ser posible. En ambos casos obtendremos un __mensaje de error__.

Los mensajes de error pueden llegar a ser muy extensos, y a veces entregan un volumen de información técnica abrumadora, pero por lo general obtendrá, __al final del mensaje__, un resumen del error que un ser humano sin tanto entrenamiento puede comprender. Este mensaje le dirá en inglés qué fue lo que ocurrió mal, y posiblemente le indique incluso cómo corregirlo.

In [18]:
3 / 0 # Esta línea lanzará un error ya que estamos dividiendo por cero!!!

Traceback (most recent call last):
  Cell In[18], line 1
    3 / 0 # Esta línea lanzará un error ya que estamos dividiendo por cero!!!
ZeroDivisionError: division by zero

Use %tb to get the full traceback.


# Funciones

Una función es un objeto que toma argumentos, y en base a ellos realiza alguna acción o retorna algún valor.

Por ejemplo, la función `pow` toma dos números `a` y `b`, y retorna `a` elevado a `b`.

In [19]:
pow(2, 3) # También lo podemos escribir como 2**3

8

Otro ejemplo de función es `print`, que toma uno o más argumentos y los imprime en pantalla.

In [20]:
print(2 ** 3)
print(2 ** 0.5) # Esto es la raíz cuadrada de 2, aproximadamente 1.4142...
print('El sentido de la vida es', 42)

8
1.4142135623730951
El sentido de la vida es 42


Notemos que una función viene identificada sólo por su nombre, y que para referirnos a ella __no le ponemos paréntesis__. Por ejemplo, nos referirnos a la función `pow` y __no__ a `pow()`.

In [21]:
pow # Esto es una función, y Python la identifica como tal

<function pow(base, exp, mod=None)>

In [22]:
# Esto falla, porque le estamos pidiendo a Python que evalúe 'pow' sin ningún argumento,
# lo que no está definido
pow()

Traceback (most recent call last):
  Cell In[22], line 3
    pow()
TypeError: pow() missing required argument 'base' (pos 1)

Use %tb to get the full traceback.


__Tip:__ Siempre que en Python vea un nombre con un paréntesis redondo a su derecha, es porque se llama a una función.

Por ejemplo:
```python
print()
int(2.5)
```
    

__Tip:__ Existe una función que le permite obtener información sobre otros objetos en Python. Esta función se llama `help` y puede, por ejemplo, darle información sobre qué hacen otras funciones y cómo utilizarlas.

In [23]:
help(pow)

Help on built-in function pow in module builtins:

pow(base, exp, mod=None)
    Equivalent to base**exp with 2 arguments or base**exp % mod with 3 arguments
    
    Some types, such as ints, are able to use a more efficient algorithm when
    invoked using the three argument form.



## Definiendo nuevas funciones

Como vimos, existen funciones intrínsecas de Python, lo que significa que vienen pre-definidas. De todos modos, es lógico que queramos definir nuestras propias funciones para extender la funcionalidad de Python, lo cual es muy sencillo de hacer.

Para definir una nueva función se utiliza la palabra `def`, y se utiliza como se muestra a continuación

In [24]:
def saludar(argumento):
    # Notar que despues del nombre de la funcion vienen los argumentos, y finaliza la declaracion con ':'
    # Aquí escribir todo lo que quiera que la función haga
    print('Hola,', argumento, '!')
    print('Es un gusto tenerte aquí :)')
    
# Aquí podría escribir más código, el cual no pertenece a la función.


Es importante notar que Python detecta qué líneas pertenecen al cuerpo de la función en base a la indentación, es decir, de acuerdo a la cantidad de espacios en blanco que tenga la línea al principio.

Esto es __muy importante__. Python espera que en un bloque de código, todas las líneas comiencen con la misma cantidad de espacios al inicio. Si dos líneas tienen distinta indentación, Python asumirá que corresponden a bloques distintos, y en este caso, terminará la definición de la función.


Ahora puede ejecutar la funcion definida como es usual

In [25]:
nombre_usuario = 'Francisco' # Cambie Francisco por su nombre aquí

saludar(nombre_usuario)

Hola, Francisco !
Es un gusto tenerte aquí :)


También es posible definir funciones con un mayor numero de argumentos, o que retornen un valor en particular. Supongamos que queremos definir una función que para tres argumentos, `x`, `y`, `z`, y calcule 

$$f(x, y, z) = 2x - 3xy + z^2.$$

Esto lo podemos hacer con la palabra `return`

In [26]:
def f(x, y, z):
    return 2*x - 3*x*y + z**2

# Probamos la función
f(1, 2, 3) # Debe dar como resultado 5

5

En el caso anterior pasamos los valores `x=1`, `y=2` y `z=3` por el orden en que estaban definidos, pero también podemos pasar estos valores en otro orden si damos su nombre

In [27]:
f(z=3, x=1, y=2) # equivalente a f(1, 2, 3)

5

Además, es posible definir las funciones con argumentos que tengan un valor por defecto, pero que puede ser especificado si el usuario lo desea

In [28]:
def f(x, y, z=0):
    return 2*x - 3*x*y + z**2

print(f(1, 2))    # Asume z=0, debe dar como resultado -4
print(f(1, 2, 3)) # Toma  z=3, retornando 5 como antes.

-4
5


### Reducción de argumentos

Es recurrente encontrarnos en situaciones en que nos interesa estudiar la dependencia de una función en sólo un subconjunto de sus argumentos, fijando el resto.

Imaginemos que queremos entregar una función `g` que haga lo mismo que la función `f` previamente definida, pero que sólo dependa de la variable `x`, fijando `y=0` y `z=0`.
¿Cómo lo podríamos implementar?

#### Ejercicio
Defina una función `g`, que tome `x`, y retorne `f(x, 0, 0)`.

### Funciones lambda

Las funciones anónimas son funciones que no tienen un identificador (nombre) asociado, y en general están pensadas para ser usadas sólo una vez y luego descartadas.

En Python, las funciones anónimas son comúmmente llamadas "funciones lambda", y proveen una sintáxis muy cómoda para definir funciones breves.

A modo de ejemplo, la instrucción
```python
lambda x: x**2 - 3
```
genera una función anónima, la cual toma un argumento, $x$, y retorna $x^2 - 3$.


También es posible generar una función que tome dos o mas argumentos. Por ejemplo, la función `f` que definimos mas arriba, que tomaba $x$, $y$ y $z$, y retornaba $2x + 3xy + z^2$, podría ser definida mas brevemente como

```python
f = lambda x, y, z: 2*x + 3*x*y + z**2
```

En particular, es muy cómodo utilizar funciones lambda para fijar argumentos de funciones previamente existentes.

In [29]:
def g(x):
    return f(x,0,0)

# Escriba su solución aquí

In [30]:
g(1)
f(1,0,0)

2

#### Ejercicio
Utilice sintaxis de funciones lambda para definir la función `g` del ejercicio anterior.

In [31]:
# Escriba su solución aquí

# Listas y Diccionarios

## Listas

Entre los distintos tipos de objeto que Python nos permite utilizar, están las listas. Estas funcionan como un contenedor __unidimensional ordenado__ que nos permite almacenar muchos valores de cualquier tipo, y se declaran utilizando paréntesis cuadrados, `[]`.

Por ejemplo, podemos definir una lista de números

In [32]:
# Podemos definir una lista de 4 elementos
mi_lista = [1.0, 5j, 'palabra', 0]

y luego podemos acceder a sus elementos indexandola con paréntesis cuadrados, con el número del elemento que queremos extraer,

In [33]:
mi_lista[2]

'palabra'

o modificar el valor guardado mediante el operador de asignación `=`,

In [34]:
mi_lista[2] = 'otra cosa'

mi_lista

[1.0, 5j, 'otra cosa', 0]

### ¡Muy importante!

¿Notó que indexamos con el número 2, pero accedimos y modificamos sobre el tercer elemento? Esto es porque __en Python siempre se cuenta desde el cero__. Es decir, que si se tienen $N$ elementos, el primero será indexado con el $0$, y el último con $N-1$.

In [35]:
print(mi_lista[0]) # 1.0
print(mi_lista[1]) # 5j
print(mi_lista[2]) # 'otra cosa'
print(mi_lista[3]) # 0

1.0
5j
otra cosa
0


y, si se intenta acceder a un elemento que no está definido, se obtendrá un error

In [36]:
mi_lista[4] # Error. El índice '4' apunta al quinto elemento, que no existe

Traceback (most recent call last):
  Cell In[36], line 1
    mi_lista[4] # Error. El índice '4' apunta al quinto elemento, que no existe
IndexError: list index out of range

Use %tb to get the full traceback.


Aunque en Python se admite entregar indices negativos para contar desde el final hacia el inicio

In [37]:
print(mi_lista[-1]) # Último elemento
print(mi_lista[-2]) # Penúltimo elemento

0
otra cosa


## Diccionarios

Un diccionario es una estructura de datos similar a una lista, pero tiene la particularidad de que, en vez de ser indexado mediante números, lo es mediante palabras.
Un diccionario se crea utilizando llaves `{}`.

In [38]:
# Los valores a la izquierda del ':' son las llaves (keys) con la cual se indexa,
# y los valores a la derecha son los valores a guardar (values).
mi_dict = {'hola'   : 10,
           'pato'   :  5,
           'nombre' : 'Loreto'}


print(mi_dict['pato'])
print(mi_dict['nombre'])

5
Loreto


Podemos crear una nueva entrada en el diccionario, o cambiar el valor de una entrada mediante el operador de asignación `=`,

In [39]:
mi_dict['telefono'] = 912345678
mi_dict['nombre'] = 'Juan'

mi_dict

{'hola': 10, 'pato': 5, 'nombre': 'Juan', 'telefono': 912345678}

# Módulos

En Python, así como muchos otros lenguajes, existe el concepto de módulos. Estos son piezas de código desarrolladas por terceras personas o grupos, y que ofrecen extender la funcionalidad del lenguaje. 

Las librerías se pueden importar ejecutando
```python
import nombre_libreria
```
lo que deja todas los objetos de la librería a disposición del usuario. Sin embargo, en general se recomienda importar las librerías bajo un alias para evitar posibles conflictos entre objetos que provengan de diferentes módulos pero tengan el mismo nombre. Por ejemplo, podemos importar la famosa librería __numpy__ bajo el alias __np__

```python
import numpy as np
```
de modo que ahora podemos acceder a las funciones u otros objetos definidos por `numpy` mediante su alias, como
```python
np.nombre_objeto
```

In [40]:
# Importamos numpy bajo el alias np
import numpy as np

# Utilizamos la función 'sqrt' de numpy para calcular raíz cuadrada
np.sqrt(2) # 1.4142135...

1.4142135623730951

También es posible importar un objeto o conjunto de objetos desde una librería, sin utilizar un alias

In [41]:
from numpy import pi
from numpy import sin, cos

print('pi =', pi)           #  3.141592...
print('sin(pi) =', sin(pi)) #  0.0
print('cos(pi) =', cos(pi)) # -1.0

pi = 3.141592653589793
sin(pi) = 1.2246467991473532e-16
cos(pi) = -1.0


Notemos que, aunque $\sin(\pi)$ es estrictamente cero, Python nos retornó el valor `1.2246467991473532e-16`, que se lee como $1.2246\ldots \times 10^{-16} = 0.00000000000000012246\ldots$.

Lo anterior ocurre porque un computador no puede almacenar un número con infinitos dígitos, y por lo tanto al trabajar con números reales (o complejos) en un computador siempre existirá error de redondeo, independiente del lenguaje que se utilice.

Aunque el error de redondeo debe ser tomado en cuenta para realizar un trabajo más elaborado, por ahora no nos preocuparemos por el, y nos quedaremos con que nuestro resultado fue correcto hasta el quinceavo ($15^\circ$) decimal, lo cual es suficiente para la mayor parte de los propósitos :).

Como se pudo observar, el paquete numpy tiene incorporado algunas funciones tal como sin(), cos(), etc.
Una función que nos será útil llamar desde numpy es la función exponencial:
\begin{equation}
e^{x}
\end{equation}
    

In [42]:
import numpy as np
np.exp(0)

1.0

## Arreglos de Numpy

Un arreglo es una estructura de datos similar a una lista, pero tiene dos características importantes:
* Puede almacenar datos en una estructura en más de una dimensión. (d=1 es vector, d=2 es matriz, etc...)
* Si almacena valores de un mismo tipo (numeros complejos, por ejemplo), es mucho más eficiente para operar que una lista.


En particular, `numpy` implementa arreglos que son mucho más eficientes que las estructuras por defecto de Python, y además permite operarlos con mucha facilidad para realizar cálculo numérico.

## Vectores en numpy


La interfaz por defecto para crear arreglos en `numpy` es la función `array`, que toma una lista y la convierte en un arreglo de `numpy`.

In [43]:
mi_vector = np.array([1,2,3])

mi_vector

array([1, 2, 3])

Aquí definimos el vector (arreglo en una dimensión)

$$ \text{mi_vector} = \begin{pmatrix} 1 \\ 2 \\ 3 \end{pmatrix} $$

que ahora podemos operar matemáticamente.

## Operaciones sobre arreglos de numpy

Podemos sumar (`+`), restar (`-`), multiplicar (`*`) y dividir (`/`) sobre arreglos de `numpy` y/o números escalares a nuestro antojo, siempre y cuando al operar sobre dos arreglos estos tengan el mismo tamaño

In [44]:
nuevo_vector = 3*mi_vector + [1, 2, 1] # Al operar un array con una lista, retorna un array

nuevo_vector

array([ 4,  8, 10])

In [45]:
nuevo_vector * [2, 3, 2] / [1, 8, 2]

array([ 8.,  3., 10.])

pero es importante notar que __estas operaciones se hacen elemento a elemento__. Es decir que al hacer el producto (`*`) de dos arreglos, el resultado es un arreglo tal que la primera componente es el producto de las primeras componentes, la segunda componente es el producto de las segundas componentes, y así...

In [46]:
np.array([1, 2, 3]) * np.array([4, 5, 6])

array([ 4, 10, 18])

lo que __no coincide con la definición de producto entre vectores__. Lo mismo aplica para cuando hacemos una división (`/`) entre dos arreglos; puede respirar con tranquilidad, nadie está dividiendo entre vectores :).

### Definición de producto entre vectores

Para hacer un producto entre vectores, `numpy` tiene la función `vdot`, que toma el complejo conjugado del primer vector, luego hace el producto por componentes entre el arreglo resultante y el segundo vector, y finalmente suma todas las componentes

In [47]:
vector1 = np.array([1j, 0, 1])
vector2 = np.array([1j, 1, 0])

np.vdot(vector1, vector2)

(1+0j)

In [48]:
# Lo anterior es equivalente a hacer
np.sum(vector1.conj() * vector2)

(1+0j)

Notamos que para tomar el complejo conjugado del primer vector escribiremos `vector1.conj()`. Esto es porque los arreglos de `numpy` traen algunas operaciones comunes definidas por defecto, como por ejemplo

* `arreglo.conj()` para tomar el complejo conjugado de `arreglo`
* `arreglo.T` para tomar la trasposición de `arreglo`
* `arreglo.real` para tomar la parte real de `arreglo`
* `arreglo.imag` para tomar la parte imaginaria de `arreglo`

## Matrices en numpy

Podemos utilizar la funcion `array` de numpy para construir una matriz a partir de una lista de listas. Es decir, entregamos a `array` una lista donde cada elemento es otra lista correspondiendo a las filas de la matriz.

Por ejemplo, para definir la matriz $\quad\begin{pmatrix} 1 & 2 & 3 \\ 4 & 5 & 6 \\ 7 & 8 & 9 \end{pmatrix}\quad$ escribimos

In [49]:
mi_matriz = np.array([[1, 2, 3],
                      [4, 5, 6],
                      [7, 8, 9]])

mi_matriz

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

y tal como antes podemos hacer operaciones componente a componente con las operaciones usuales (`+`, `-`, `*`, `/`)

In [50]:
mi_matriz * mi_matriz # matriz correspondiente al cuadrado de cada elemento de 'mi_matriz'

array([[ 1,  4,  9],
       [16, 25, 36],
       [49, 64, 81]])

Podemos crear una matriz identidad de $3\times3$ a mano, o mediante la función `eye` de `numpy`,

In [51]:
identidad = np.eye(3)

identidad

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

y utilizarla para notar que, tal como en el caso anterior de vectores, el producto usual no calza con la definición de producto entre matrices

In [52]:
identidad * mi_matriz

array([[1., 0., 0.],
       [0., 5., 0.],
       [0., 0., 9.]])

Para el producto entre matrices, `numpy` reserva el símbolo `@`

In [53]:
identidad @ mi_matriz # Debe retornar lo mismo que mi_matriz

array([[1., 2., 3.],
       [4., 5., 6.],
       [7., 8., 9.]])

#### Ejercicio

Defina una función llamada `semiposdef` que tome como argumento una matriz $A$ y retorne el producto $A^{\dagger}A$, donde $A^{\dagger}$ es la matriz $A$ traspuesta y complejo-conjugada.

In [54]:
# Escriba su solución aquí

# Control de flujo

Hasta el momento, salvo por el uso de funciones, sólo hemos visto código que se ejecuta de forma secuencial y lineal. Sin embargo, es posible controlar el orden de ejecución del código separandolo en bloques y utilizando estructuras de control de flujo. En particular estudiaremos dos tipos de estas estructuras

## Ejecución condicional

Permiten la ejecución de bloques de código cuando se verifica una condición. La más básica de estas estructuras se escribe

```python
if condicion:
    bloque
    de
    codigo
```

donde `condicion` es una variable lógica que, de ser verdadera, se ejecuta el bloque de código.

Una extensión de esta estructura es la llamada `if-else`,

```python
if condicion:
    primer
    bloque
else:
    segundo
    bloque
```

y se lee: "Sí `condicion`, entonces ejecuta el primer bloque, sino ejecuta el segundo".

In [55]:
a = 3

if a > 5:
    print('a es mayor que 5')
else:
    print('a es menor o igual a 5')

a es menor o igual a 5


## Ejecución iterativa

Estructuras que permiten ejecutar un bloque de código repetidas veces. La más común es el ciclo `for`, que se escribe

```python
for variable in coleccion:
    bloque
    de
    codigo
```

y se lee "para cada valor de `variable` en `colección`, ejecuta el bloque de código", donde `coleccion` es una variable que contiene un conjunto de elementos, como puede ser una lista o arreglo.

In [56]:
for i in [1,2,3,4,5]:
    i2 = i*i
    print('i^2 =', i2)
print('Esta línea tiene otra indentación así que pertenece a otro bloque. Está fuera del ciclo.')

i^2 = 1
i^2 = 4
i^2 = 9
i^2 = 16
i^2 = 25
Esta línea tiene otra indentación así que pertenece a otro bloque. Está fuera del ciclo.


Una objeto que comunmente se utiliza en Python para ejecución iterativa es el iterador `range`, que dado como argumento un número entero $N$, retorna los enteros desde $0$ hasta $N-1$.

In [57]:
for i in range(10):
    print('i =', i)

i = 0
i = 1
i = 2
i = 3
i = 4
i = 5
i = 6
i = 7
i = 8
i = 9


Ejemplo: Dado que los arreglos de `numpy` proveen su tamaño mediante su propiedad `size`, uno podría mostrar todos los elementos de un vector como

In [58]:
arreglo = np.array([1, 4, 7, 9])

for i in range(arreglo.size):
    print('arreglo[', i, '] =', arreglo[i])    

arreglo[ 0 ] = 1
arreglo[ 1 ] = 4
arreglo[ 2 ] = 7
arreglo[ 3 ] = 9


#### Ejercicio

Sabiendo que puede repetir un _string_ multiplicándolo por un entero,
```python
print(5*'-') # Imprime -----
```
Utilice ciclos `for` y la función `print` para imprimir en pantalla la figura

```
*
**
***
****
*****
```

In [59]:
# Escriba su solución aquí

#### Ejercicio

Similar al ejercicio anterior, escriba una pieza de código que imprima la figura
```
*
**
***
****
*****
****
***
**
*
```

In [60]:
# Escriba su solución aquí

#### Ejercicio

Escriba una función llamada `triangulo`, que tome un argumento `N`, e imprima en pantalla
la misma figura del ejercicio anterior, pero generalizada tal que la fila central tiene largo `N`.

In [61]:
# Escriba su solución aquí

### Ciclos implícitos

Es posible crear colecciones, como listas, de tamaño arbitrariamente largo utilizando notación de ciclo implícito, donde se evalúa una expresión para cada elemento de otra colección. Un ejemplo de esto es

In [62]:
mi_lista = [2*i for i in range(10)]

mi_lista

[0, 2, 4, 6, 8, 10, 12, 14, 16, 18]

#### Ejercicio
Utilice notacion de ciclo implicito para definir una lista cuyos elementos sean los elementos de 'mi_lista' pero elevados al cuadrado.

In [63]:
# Escriba su solución aquí

## Clases, objetos y atributos

Es común oír hablar de clases y objetos. Si bien no está en el objetivo de este tutorial aprender a definirlos a nuestro antojo, sí nos resultará útil aprender, a _grosso modo_, que significan estos conceptos.

Podemos entender los objetos como entes con características dadas, los cuales podemos utilizar y/o manipular. En este contexto, las clases corresponderían a un tipo de objetos, los cuales tienen un conjunto de características en común.

Un ejemplo típico que se utiliza para aterrizar estos conceptos es el siguiente. Consideremos una clase a la que llamamos `Perro`, y caractericémosla de manera que todos sus elementos tengan atributos de `nombre` y `color`.
Dada esta clase, podemos definir un objeto o instancia en particular de `Perro` tal que su nombre sea 'Firulais' y su color 'Café'.

En Python, practicamente todo es un objeto, y podemos acceder a sus atributos mediante la sintaxis
```python
objeto.atributo
```

¿Recuerda que mas atrás accedimos a ciertas funcionalidades de los arreglos de `numpy` con esta sintaxis?

Por ejemplo, cuando teníamos un vector complejo,

In [64]:
v = np.array([0, 0, 1j])
v

array([0.+0.j, 0.+0.j, 0.+1.j])

y dijimos que podíamos obtener sus partes imaginarias como `v.imag`

In [65]:
v.imag

array([0., 0., 1.])

Esto es porque los objetos del tipo `numpy.ndarray` (array de numpy), tienen un atributo `imag` que corresponde al arreglo formado por las partes imaginarias de cada una de sus componentes.

### Métodos

Tal como un objeto tiene atributos, que son características que lo definen, también puede tener métodos asociados. Estos métodos corresponden a acciones (funciones) que están asociadas al objeto en cuestión.

Por ejemplo, los arrays de `numpy` tienen el método `conj`, de modo que podemos obtener el vector complejo-conjugado de `v` como

In [66]:
v.conj()

array([0.-0.j, 0.-0.j, 0.-1.j])

Particularmente interesante es que, si bien en el caso de `numpy` estos métodos sólo leen información del objeto sin cambiarlo, en general sí es posible aplicar modificaciones a un objeto mediante sus métodos, lo cual haremos bastante durante los próximos dias.

__Tip:__ Cuando está trabajando en un Jupyter Notebook, puede escribir el nombre de un objeto seguido de un punto `.`, y presionar la tecla `<Tab>` (la que tiene el símbolo ⭾, al costado izquierdo de su teclado) para mostrar una lista con los atributos y métodos del objeto en cuestión.

```
objeto.<Tab>
```

## Contactos: 

Jorge Gidi <br> 
Estudiante doctorado en ciencias físicas <br> 
Universidad de Concepción <br>
jorgegidi@udec.cl

Francisco Jara <br>
Estudiante magíster en ciencias físicas <br>
Universidad de Concepción <br>
Frjara2019@udec.cl

In [1]:
## Contactos: 

import qiskit.tools.jupyter
%qiskit_version_table

c:\programdata\miniconda3\lib\site-packages\numpy\_distributor_init.py:30: UserWarning: loaded more than 1 DLL from .libs:
c:\programdata\miniconda3\lib\site-packages\numpy\.libs\libopenblas.FB5AE2TYXYH2IJRDKGDGQ3XBKLKTF43H.gfortran-win_amd64.dll
c:\programdata\miniconda3\lib\site-packages\numpy\.libs\libopenblas.JPIJNSWNNAN3CE6LLI5FWSPHUT2VXMTH.gfortran-win_amd64.dll
  warnings.warn("loaded more than 1 DLL from .libs:"
